evalml/pipelines/components/transformers/preprocessing/time_series_regularizer.py

```diff
@@ -0,0 +1,336 @@
+    on the estimated frequency for `time_index`. In the original X and y passed:
+     - Missing datetime values will be added and will have their corresponding columns in X and y set to None.
+     - Duplicate datetime values will be dropped.
+     - Extra datetime values will be dropped.
+     - If it can be determined that a duplicate or extra value is misaligned, then it will be repositioned to take the
+     place of a missing value.
+
+    This Transformer should be used before the `TimeSeriesImputer` in order to impute the missing values that were
+    added to X and y (if passed).
+
+    Args:
+        time_index (string): Name of the column containing the datetime information used to order the data.
+        random_seed (int): Seed for the random number generator. This transformer performs the same regardless of the random seed provided.
+    """
+
+    name = "Time Series Regularizer"
+    hyperparameter_ranges = {}
+    """{}"""
+
+    modifies_target = True
+    training_only = True
+
+    def __init__(
+        self, time_index=None, window_length=5, threshold=0.8, random_seed=0, **kwargs
+    ):
+        self.time_index = time_index
+        self.window_length = window_length
+        self.threshold = threshold
+        self.error_dict = {}
+        self.inferred_freq = None
+        self.debug_payload = None
+
+        parameters = {
+            "time_index": time_index,
+            "window_length": window_length,
+            "threshold": threshold,
+        }
+        parameters.update(kwargs)
+
+        super().__init__(parameters=parameters, random_seed=random_seed)
+
+    def fit(self, X, y=None):
+        """Fits the TimeSeriesRegularizer.
+
+        Args:
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features].
+            y (pd.Series, optional): The target training data of length [n_samples].
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: if self.time_index is None
+            TypeError: if the `time_index` column is not of type Datetime
+            ValueError: if X and y have different lengths
+            ValueError: if `time_index` in X does not have an offset frequency that can be estimated
+        """
+        if self.time_index is None:
+            raise ValueError("The argument time_index cannot be None!")
+
+        X_ww = infer_feature_types(X)
+
+        if not isinstance(X_ww.ww.logical_types[self.time_index], Datetime):
+            raise TypeError(
+                f"The time_index column {self.time_index} must be of type Datetime."
+            )
+
+        if y is not None:
+            y = infer_feature_types(y)
+            if len(X_ww) != len(y):
+                raise ValueError(
+                    "If y has been passed, then it must be the same length as X."
+                )
+
+        ww_payload = infer_frequency(
+            X[self.time_index],
+            debug=True,
+            window_length=self.window_length,
```

Comment:
Initially the WW PR had this set to 5 and 0.8 respectively and then Dave intended to increase the window size to 15. He and I spoke about this and how this might affect smaller datasets (which are more prone to not having an offset frequency detected by WW inference logic if the window size and threshold are too high).

So Dave made them parameters that could be changed and I decided upon 5 and 0.8 which were values that allowed for a small handful (4-5 usually) of faulty datetime values for a dataset as small as 100 observations. If they were set to 15 and 0.9, a dataset that small wouldn't be allowed even a single faulty value.

I think a good V2 of this would include a smart function that changes these parameters based on the size of the dataset (if the user doesn't pass in their own values).

Reply:
@ParthivNaresh can you file this followup story for me so we don't forget

Reply:
@chukarsten Done, the issue has been filed [here](https://github.com/alteryx/evalml/issues/3441)!